### DISCUSSION TITLE
Numbering overhall

### DISCUSSION THREAD

**Discussion by emilyyyylime** [typst/typst#2353] [Ideas]
I think what has been done with `datetime.display`'s format string is very good, as it's less ambiguous and more flexible than `numbering`'s dated format.

It's better to break some code now than to stay stuck with a fragile and restrictive system forever; we could even make it backwards compatible for most 'good' cases, though that might be a bit difficult.

I hope this thread can serve as a place to discuss possible alternatives to the current system, and eventually lead to it being replaced.
I also especially want feedback from speakers of other languages and generallu users as to what use-cases you've had to implement with functions manually.

Some unfortunate cases to consider:
- `set page(numbering: "Page 1")` is interpreted as literally "P", the page number as a lexical ordinal (a, b, c, ...), literally "ge ", then the total page count as an Arabic numeral.
- No way to repeat from the start when done (i.e. you might want the pattern `"1.a.i."` with five levels to resolve as `"1.a.i.1.a."` rather than `"1.a.i.i.i."`).
- No way to differentiate different levels of nesting on their own — for example to have `enum`s on the top level use `1.` and those below them use `a.`, currently you either must use some weird `show` and `set` shenanigans, or define a custom function with `full: true`. This can be solved in numbering if there were an easy way to specify only the last part is to be used.
- No way to specify variants of the same numbering, for example having a component padded to a certain width (for example for numbering code blocks). Another example is Hebrew numbering with `א` without 'Geresh' and/or 'Gershayim', or starting Arabic numerals from 0.

These are all very common use cases and I think they should be as convenient as current well-supported use cases are.

**Comment by emilyyyylime** [comment]
Either way I think changes will be have to be made to the way the `numbering` function reads its pattern argument(s?), and additional switches should be added to it.
The general direction I was thinking we should pursue would be along the direction of `numbering("Verbatim text [component(:fmt)] more text", full: bool, last-only: bool, repeat: bool)`, or maybe something like `numbering("Verbatim text ", "component(:fmt)", " more text")`. 
With both a consideration to have is how to specify a different separator based on whether a component is the last one (current `1.1)` behaviour, which I would not want to lose).

**Comment by PgBiel** [comment]
See also: https://github.com/typst/packages/tree/main/packages/preview/numberingx

**Comment by edhebi** [comment]
Fwiw, numberingx reimplement the exact same rules as `numbering()` by design, so it also inherits their limitations. For the "counters" themselves, I do think reusing the web spec is good, it both makes it a bit more documented/discoverable, and it also honestly provides solid internationalization.

In general numbering patterns shouldn't have a ton of text or decoration attached tbh, that should be the job of a show rule on the component with that numbering, otherwise you get weird results with references or outlines.

On particular usecase that is currently very manual and verbose is being able to use different patterns based on level without repeating the previous ones. I'm not entirely sure what the best ergonomics would be

**Comment by tingerrr** [comment]
> In general numbering patterns shouldn't have a ton of text or decoration attached tbh, that should be the job of a show rule on the component with that numbering, otherwise you get weird results with references or outlines.

@laurmaedje has mentioned that numbering should be an inherent property of a counter, and therefor the show rule for a counter would define those non numbering decorations like `(prefix: "Page ", "1")`. But this would require them to know where and how they are displayed. Think of elements which may display their own supplement with their counter/numbering, leading to referencing having 2 supplements and similar issues.

> On particular usecase that is currently very manual and verbose is being able to use different patterns based on level without repeating the previous ones. I'm not entirely sure what the best ergonomics would be

@Dherse has mentioned on discord that a specific type for numberings could be used to make the creation of these repeating patterns easier without having to drop into the `(..args) => ...` escape hatch. Relevant discussions are here on the [community discord](https://discord.com/invite/2uDybryKPe):
- initial: https://discord.com/channels/1054443721975922748/1160023808791543869/1160152679096520704
- continuation in: https://discord.com/channels/1054443721975922748/1160163393332858981

As can be seen in the discussions linked above, counters/numberings which have semantic relationships to their location, are currently doable but require packages which either rely on reimplementing show rules for outlines (anti-matter) or need to manually specify all figure kinds (i-figured) to ensure the correct resets.

Often times the location is all you need to get the correct numbering. @laurmaedje (see second link) mentioned that this could be improved once [contexts](https://laurmaedje.github.io/posts/types-and-context/) can be provided to functions, so I think the semantic relationship problem is partially solved if contexts land.
